src/Plugin/Autoloader.php

```diff
@@ -0,0 +1,219 @@
+use Kirby\Exception\Exception;
+use Kirby\Exception\InvalidArgumentException;
+use Kirby\Filesystem\F;
+use Kirby\Toolkit\Str;
+use Throwable;
+
+class Autoloader
+{
+
+    public array $extends = [];
+
+    /**
+     * @param string $name Plugin name
+     * @param string $root The root path of the plugin
+     * @param array $data Predefined extension data to extend
+     * @return void|$this 
+     */
+    public function __construct(
+        public string $name,
+        public string $root,
+        public array $data = []
+    ) {
+
+        //Load classes before everything else
+        $classfolder = $root . '/classes/';
+        if (Dir::exists($classfolder)) {
+            $this->_classes($classfolder);
+        }
+
+        foreach (Dir::dirs($this->root) as $dir) {
+            if (method_exists($this, $dir)) {
+                $this->$dir($this->root . '/' . $dir . '/');
+            }
+        }
+    }
+
+    /**
+     * Walk throuh the given path
+     * @param string $root Absolute path to walk
+     * @param Closure $fnc Callback for each file
+     * @return void 
+     */
+    public function _dirWalker(string $root, Closure $fnc)
+    {
+
+        foreach (Dir::index($root, true) as $path) {
+
+            //Check if the path is active
+            if (Str::contains($path, '/_')) {
+                continue;
+            }
+
+            $file = $root . $path;
+            if (F::exists($file)) {
+                $dirname = F::dirname($path);
+                $dirname = $dirname === '.' ? '' : $dirname . '/';
+                $fnc($dirname . F::name($path), $file);
+            }
+        }
+    }
+
+    /**
+     * Read file by given name and check if available
+     * @param string $file 
+     * @return array 
+     * @throws Throwable 
+     * @throws InvalidArgumentException 
+     * @throws Exception 
+     */
+    public function _read(string $file): array
+    {
+
+        $content = Data::read($file);
+
+        //Value may be a closure
+        if (is_array($content) && count($content) === 0) {
+            $name = F::relativepath($file, $this->root);
+            throw new Exception("The content for '$name' cannot be resolved by the autoloader");
+        }
+
+        return $content;
+    }
+
+    /**
+     * Autoload autoload
+     * @param string $root 
+     * @return void 
+     */
+    public function autoload(string $root)
+    {
+        $this->_dirWalker($root, function ($path, $file) {
+            $this->extends[$path] = $this->_read($file);
+        });
+    }
+
+    /**
+     * Autoload blueprints
+     * @param string $root 
+     * @return void 
+     */
+    public function blueprints(string $root)
+    {
+        $this->_dirWalker($root, function ($path, $file) {
+
+            $name = F::relativepath($file);
+            //Set YAML-File or they content
+            $this->extends['blueprints'][$path] = (F::extension($file) === 'yml') ? $file : $this->_read($file);
+        });
+    }
```

Comment:
As an example, I think this could be refactored more in this direction:

```suggestion
    public function loadBlueprints(): void
    {
        $root = $this->root . '/blueprints';
        
        foreach (Dir::index($root, true) as $blueprint) {
            $file = $root . '/' . $blueprint;
            
            if (F::exists($file) === false) {
                continue;
            }
            
            $this->extends['blueprints'][$blueprint] =  $$file;
        }
    }
```

(Just writing this on GitHub so code is probably not fully right, but to communicate the general idea)

Reply:
A good point. Would that mean that an array containing all specific folders with methods? For example, with:

```
$folders = [
    'snippets' => 'loadSnippets'
    ...
];

foreach (Dir::dirs($this->root) as $dir) {
    if (array_key_exists($dir, $folders) === false) {
        $method = $folders[$dir];
        $this->$method($this->root . '/' . $dir . '/');
    }
}
´´´

Reply:
I would not try to make it that smart/magic. I know it seems very verbose and repetitive, but from our experience "boring" code like that has often paid off down the road.